### PR TITLE
[VMAP] Add linspace and logspace batch rules

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105451
* #104889
* #107958

### CODE REVIEW

**Comment by zou3519** [comment]
You might need to rebase?

File: aten/src/ATen/functorch/BatchRulesFactory.cpp [hunk 1]
```diff
@@ -103,6 +103,121 @@ static std::tuple<Tensor,optional<int64_t>> _new_zeros_with_same_feature_meta_ba
...
+  start_ = moveBatchDimToFront(start_, start_bdim);
+  end_ = moveBatchDimToFront(end_, end_bdim);
+
+  auto tensor_options = at::TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory);
+
+  Tensor result;
+  if (steps == 0){
+    result = at::full({batch_size, 0}, 0, tensor_options);
+  } else if (steps == 1){
+    result = at::empty({batch_size}, tensor_options).copy_(start_).unsqueeze(1);
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchRulesFactory.cpp:128]: could we do `start_.new_empty({batch_size}, tensor_options)`? If start_ is being vmapped over I am not sure the current code would work

File: aten/src/ATen/functorch/BatchRulesFactory.cpp [hunk 2]
```diff
@@ -103,6 +103,121 @@ static std::tuple<Tensor,optional<int64_t>> _new_zeros_with_same_feature_meta_ba
   return std::make_tuple(result, 0);
 }
 
+static std::tuple<Tensor,optional<int64_t>> linspace_logspace_batch_rule_helper(
+    const at::Tensor& start, optional<int64_t> start_bdim,
+    const at::Tensor& end, optional<int64_t> end_bdim,
+    int64_t steps,
+    c10::optional<double> base,
```
[the remark this answers is not in the document]
Comment (qqaatw) [reply aten/src/ATen/functorch/BatchRulesFactory.cpp:110]: Yes.

File: aten/src/ATen/functorch/BatchRulesFactory.cpp (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/functorch/BatchRulesFactory.cpp]: Yes, the tests passed.

**Review by zou3519 — APPROVED** [review]

**Comment by qqaatw** [comment]
@pytorechbot merge -ic

**Comment by qqaatw** [comment]
@pytorechbot merge -i

**Pull request closed without merging**